core/tests/protractor.conf.js

```diff
@@ -158,7 +104,83 @@ exports.config = {
     users: [
       'protractor_desktop/userJourneys.js',
     ],
+  };
+
+// A reference configuration file.
+exports.config = {
+  // ----- How to setup Selenium -----
+  //
+  // There are three ways to specify how to use Selenium. Specify one of the
+  // following:
+  //
+  // 1. seleniumServerJar - to start Selenium Standalone locally.
+  // 2. seleniumAddress - to connect to a Selenium server which is already
+  //    running.
+  // 3. sauceUser/sauceKey - to use remote Selenium servers via SauceLabs.
+  //
+  // If the chromeOnly option is specified, no Selenium server will be started,
+  // and chromeDriver will be used directly (from the location specified in
+  // chromeDriver)
+
+  // The location of the selenium standalone server .jar file, relative
+  // to the location of this config. If no other method of starting selenium
+  // is found, this will default to
+  // node_modules/protractor/selenium/selenium-server...
+  seleniumServerJar: null,
+  // The port to start the selenium server on, or null if the server should
+  // find its own unused port.
+  seleniumPort: null,
+  // Chromedriver location is used to help the selenium standalone server
+  // find chromedriver. This will be passed to the selenium jar as
+  // the system property webdriver.chrome.driver. If null, selenium will
+  // attempt to find chromedriver using PATH.
+  chromeDriver: './selenium/chromedriver',
+  // If true, only chromedriver will be started, not a standalone selenium.
+  // Tests for browsers other than chrome will not run.
+  chromeOnly: false,
+  // Additional command line options to pass to selenium. For example,
+  // if you need to change the browser timeout, use
+  // seleniumArgs: ['-browserTimeout=60'],
+  seleniumArgs: [],
+  beforeLaunch: function() {
+    Object.keys(suites).forEach(function(key) {
+      var patterns = suites[key];
+      for (var pattern of patterns) {
+        var fullPattern = path.resolve(__dirname, pattern);
+        var files = glob.sync(fullPattern);
+        if (files.length == 0) {
+          var errorMessage = '';
+          if (glob.hasMagic(pattern)) {
```

Comment:
I don't understand why this is necessary. As I understand, hasMagic returns true if the "pattern" has a special character. But why do we need to check that?

Reply:
Just to have a specific message, if it doesn't have a pattern then it's a file.. so we can say there "There isn't such file."